test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -694,19 +702,17 @@ private String buildNodeName(int id, Settings settings) {
      * returns a suffix string based on the node role. If no explicit role is defined, the suffix will be empty
      */
     private static String getRoleSuffix(Settings settings) {
-        String suffix = "";
-        if (Node.NODE_MASTER_SETTING.exists(settings) && Node.NODE_MASTER_SETTING.get(settings)) {
-            suffix = suffix + DiscoveryNodeRole.MASTER_ROLE.roleNameAbbreviation();
+        if (NodeRoleSettings.NODE_ROLES_SETTING.exists(settings) == false) {
+            return "";
         }
-        if (Node.NODE_DATA_SETTING.exists(settings) && Node.NODE_DATA_SETTING.get(settings)) {
-            suffix = suffix + DiscoveryNodeRole.DATA_ROLE.roleNameAbbreviation();
-        }
-        if (Node.NODE_MASTER_SETTING.exists(settings) && Node.NODE_MASTER_SETTING.get(settings) == false &&
-            Node.NODE_DATA_SETTING.exists(settings) && Node.NODE_DATA_SETTING.get(settings) == false
-            ) {
-            suffix = suffix + "c";
+
+        final List<DiscoveryNodeRole> roles = NodeRoleSettings.NODE_ROLES_SETTING.get(settings);
+
+        if (roles.isEmpty()) {
+            return "c";
         }
-        return suffix;
+
+        return roles.stream().sorted().map(DiscoveryNodeRole::roleNameAbbreviation).collect(Collectors.joining());
```

Comment:
This is super-nit-picky, but does it make sense to map and then sort, in case the abbreviation, for some reason, ends up making the result look unsorted? It's actually what `DiscoveryNode#toString()` does too.

```suggestion
        return roles.stream().map(DiscoveryNodeRole::roleNameAbbreviation).sorted().collect(Collectors.joining());
```